### ISSUE TITLE
CA2021 Do not call Enumerable.Cast<T> or Enumerable.OfType<T> false positives

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/roslyn-analyzers#6457]
### Analyzer

**Diagnostic ID**: [CA2021](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2021): `Do not call Enumerable.Cast<T> or Enumerable.OfType<T> with incompatible types`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

### Describe the bug
Following false positives found in runtime test builds (normally we don't turn on analyzers in test projects, but this analyzer `warns` by default therefore failing the test builds). So, [we have lowered the analyzer severity](https://github.com/dotnet/roslyn-analyzers/pull/6456#pullrequestreview-1270407583) into `info` level until these bugs fixed. @fowl2 please let us know if you interested in fixing these, thanks!

1. False positive with dynamic expressions:
```cs
  // Type 'int' is incompatible with type 'dynamic' and cast attempts will throw InvalidCastException at runtime
     int[] numbers = new int[] { 1, 2, 3 };
     var v = from dynamic d in numbers select (int)d;
```
2. False positive with nullability annotations:
```cs
  // Type 'int' is incompatible with type 'int?' and cast attempts will throw InvalidCastException at runtime
    Enumerable.Range(1, 5).Cast<int?>();
```
3. False positive with ValueTuple
```cs
  // Type '(int, int)' is incompatible with type '(int min, int max)' and cast attempts will throw InvalidCastException at runtime
    from (int min, int max) pair in new[] { (1, 2), (-10, -3) };
```
4. For this one don't have a sample repro yet, check the [runtime code for more details](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/AttRegexTests.cs#L417-L420)
```cs
// `Type 'System.Collections.Generic.KeyValuePair<string, System.Text.RegularExpressions.Group>' is incompatible with type 'System.Text.RegularExpressions.Group' and cast attempts will throw InvalidCastException at runtime`
     var actualSet = new HashSet<(int start, int end)>(
                match.Groups
                .Cast<Group>()
                .Select(g => (start: g.Index, end: g.Index + g.Length)));
```

### Expected behavior
No warning
### Actual behavior
Warns
## Additional context

Found in: https://github.com/dotnet/runtime/pull/80916/commits/f3bcd214b8b604d5cd26fe6060dac659861b9039

CC @fowl2 @stephentoub

**Comment by fowl2** [comment]
@buyaa-n I've started work on this

**Comment by hoyosjs** [comment]
There's one other case I think this is throwing a warning where I didn't quite expect it:

```c#
        [Fact]
        public async Task CanCastReadonlyStructToInterface()
        {
            await VerifyCS.VerifyAnalyzerAsync(@"
using System.Collections;
using System.Collections.Generic;
using System.Linq;

interface IInterface {}

readonly struct Implementation : IInterface {}

class C
{
    void M()
    {
        IEnumerable<Implementation> e = default;
        e.Cast<IInterface>();
    }
}
");
        }
```

The problem is

https://github.com/dotnet/roslyn-analyzers/blob/main/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzer.cs#L318-L319

It doesn't consider interfaces on structs.

**Comment by buyaa-n** [comment]
> It doesn't consider interfaces on structs.

Right, @hoyosjs is this blocking any repo build or PR?

CC @fowl2 could you fix this?

**Comment by hoyosjs** [comment]
I submitted the fix PR. Its not blocking. We wanted to enable the rule in a repo that's gonna be merged into another one that has the rule enabled. This flags and breaks the build. We'll update the analyzer when it becomes available.

Cc: @leculver

**Comment by buyaa-n** [comment]
Thanks @hoyosjs, the fix merged, closing this issue
